lib/sycamore/sycamore/utils/time_trace.py

```diff
@@ -0,0 +1,61 @@
+import os
+import time
+import struct
+import resource
+import threading
+
+
+class TimeTrace:
+    fd = -1
+
+    def __init__(self, name: str):
+        self.name = name.encode()
+        self._setup()
+
+    def __enter__(self):
+        self.start()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.end()
+
+    def start(self):
```

Comment:
`# TODO: replace all of this logic with metadata once that gets in`

Normally I'd complain a bunch about how all this works, but it all becomes obsolete once metadata goes in, and we'll want metadata since that will also work on multi-node.  It's unclear to me how much of the instrumentation will remain once we have metadata. Almost all of it is whole operator instrumentation which we should do automagically.

Reply:
I'm not convinced that whole-operator is the only thing we'll want.  It's a thing we should make easy, though.  Let me know what "metadata" means and how I can integrate with it.